include/deal.II/base/index_set.h

```diff
@@ -497,6 +689,246 @@ IndexSet complete_index_set (const unsigned int N)
+{
+  Assert (accessor.index_set == other.accessor.index_set, ExcInternalError());
+
+  return static_cast<std::ptrdiff_t>(accessor.range_idx) - static_cast<std::ptrdiff_t>(other.accessor.range_idx);
+}
+
+
+/* ElementIterator */
+
+inline
+bool IndexSet::ElementIterator::is_valid() const
+{
+  return range_idx < index_set.ranges.size() && idx<index_set.ranges[range_idx].end;
+}
+
+inline
+IndexSet::ElementIterator::ElementIterator(const IndexSet &idxset, IndexSet::size_type range_idx, IndexSet::size_type index)
+  : index_set(idxset), range_idx(range_idx), idx(index)
+{
+  Assert(range_idx < index_set.ranges.size(), ExcMessage(""));
+  Assert(idx >= index_set.ranges[range_idx].begin, ExcMessage(""));
+  Assert(idx <= index_set.ranges[range_idx].end, ExcMessage(""));
+}
+
+inline
+IndexSet::ElementIterator::ElementIterator(const IndexSet &idxset)
+  : index_set(idxset), range_idx(0), idx(0)
+{
+  if (index_set.ranges.size()>0)
+    {
+      range_idx = index_set.ranges.size()-1;
+      idx = index_set.ranges[range_idx].end;
+    }
+}
+
+inline
+const IndexSet::size_type &
+IndexSet::ElementIterator::operator* () const
+{
+  Assert(range_idx < index_set.ranges.size(), ExcMessage(""));
+  return idx;
+}
+
+inline
+bool IndexSet::ElementIterator::operator == (const IndexSet::ElementIterator &other) const
+{
+  Assert (index_set == other.index_set, ExcInternalError());
+  return range_idx == other.range_idx && idx==other.idx;
+}
+
+inline
+void IndexSet::ElementIterator::advance ()
+{
+  Assert(is_valid(), ExcMessage(""));
+  if (idx < index_set.ranges[range_idx].end)
+    ++idx;
+  if (idx == index_set.ranges[range_idx].end)
+    {
+      if (range_idx < index_set.ranges.size()-1)
+        {
+          ++range_idx;
+          idx = index_set.ranges[range_idx].begin;
+        }
+    }
+}
+
+inline
+IndexSet::ElementIterator &
+IndexSet::ElementIterator::operator++ ()
+{
+  advance();
+  return *this;
+}
+
+inline
+IndexSet::ElementIterator
+IndexSet::ElementIterator::operator++ (int)
+{
+  IndexSet::ElementIterator it = *this;
+  advance();
+  return it;
+}
+
+inline
+bool IndexSet::ElementIterator::operator != (const IndexSet::ElementIterator &other) const
+{
+  return !(*this == other);
+}
+
+inline
+bool IndexSet::ElementIterator::operator < (const IndexSet::ElementIterator &other) const
+{
+  Assert (index_set == other.index_set, ExcInternalError());
+  return range_idx < other.range_idx || (range_idx == other.range_idx && idx<other.idx);
+}
+
+inline
+std::ptrdiff_t IndexSet::ElementIterator::operator - (const IndexSet::ElementIterator &other) const
+{
+  Assert (index_set == other.index_set, ExcInternalError());
+  if (*this == other)
+    return 0;
+  if (!(*this < other))
+    return -(other-*this);
+
+  std::ptrdiff_t c = -static_cast<std::ptrdiff_t>(idx-index_set.ranges[range_idx].begin);
+
+  for (size_type range=range_idx; range<other.range_idx; ++range)
+    c += index_set.ranges[range].end-index_set.ranges[range].begin;
```

Comment:
I found the negation in the first assignment to `c` surprisingly hard to read. Why not initialize `c` with the *remaining* elements in the current range and then just start the loop with `range_idx+1`? After all, the loop does not include `other.range_idx` either, so may as well not include `this->range_idx`.